In [1]:
!pip install sklearn --upgrade

Requirement already up-to-date: sklearn in c:\users\kandr\appdata\local\continuum\anaconda3\lib\site-packages (0.0)


In [2]:
!pip install joblib

In [3]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.linear_model import ElasticNet
from sklearn import tree


In [4]:
import warnings
warnings.simplefilter('ignore')

In [5]:
df = pd.read_csv("exoplanet_data.csv")
# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')
# Drop the null rows
df = df.dropna()
df.head()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 41 columns]

In [6]:
Xtemp = df[['koi_fpflag_nt', 'koi_fpflag_ss', 'koi_fpflag_co', 'koi_fpflag_ec', 'koi_period','koi_time0bk','koi_slogg','koi_srad','koi_impact','koi_duration','koi_depth','koi_prad','koi_teq','koi_insol','koi_model_snr','koi_steff','koi_slogg','koi_srad','ra','dec','koi_kepmag']]
Xtemp

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
0                 0              0              0              0   54.418383   
1                 0              1              0              0   19.899140   
2                 0              1              0              0    1.736952   
3                 0              0              0              0    2.525592   
4                 0              0              0              0    4.134435   
...             ...            ...            ...            ...         ...   
6986              0              0              0              1    8.589871   
6987              0              1              1              0    0.527699   
6988              0              0              0              0    1.739849   
6989              0              0              1              0    0.681402   
6990              0              0              1              1    4.856035   

      koi_time0bk  koi_slogg  koi_srad  koi_impact  koi_duration  ...  \
0      162.513840      4.467     0.927       0.586       4.50700  ...   
1      175.850252      4.544     0.868       0.969       1.78220  ...   
2      170.307565      4.564     0.791       1.276       2.40641  ...   
3      171.595550      4.438     1.046       0.701       1.65450  ...   
4      172.979370      4.486     0.972       0.762       3.14020  ...   
...           ...        ...       ...         ...           ...  ...   
6986   132.016100      4.296     1.088       0.765       4.80600  ...   
6987   131.705093      4.529     0.903       1.252       3.22210  ...   
6988   133.001270      4.444     1.031       0.043       3.11400  ...   
6989   132.181750      4.447     1.041       0.147       0.86500  ...   
6990   135.993300      4.385     1.193       0.134       3.07800  ...   

      koi_prad  koi_teq  koi_insol  koi_model_snr  koi_steff  koi_slogg  \
0         2.83      443       9.11           25.8       5455      4.467   
1        14.60      638      39.30           76.3       5853      4.544   
2        33.46     1395     891.96          505.6       5805      4.564   
3         2.75     1406     926.16           40.9       6031      4.438   
4         2.77     1160     427.65           40.2       6046      4.486   
...        ...      ...        ...            ...        ...        ...   
6986      1.11      929     176.40            8.4       5638      4.296   
6987     29.35     2088    4500.53          453.3       5638      4.529   
6988      0.72     1608    1585.81           10.6       6119      4.444   
6989      1.07     2218    5713.41           12.3       6173      4.447   
6990      1.05     1266     607.42            8.2       6469      4.385   

      koi_srad         ra        dec  koi_kepmag  
0        0.927  291.93423  48.141651      15.347  
1        0.868  297.00482  48.134129      15.436  
2        0.791  285.53461  48.285210      15.597  
3        1.046  288.75488  48.226200      15.509  
4        0.972  296.28613  48.224670      15.714  
...        ...        ...        ...         ...  
6986     1.088  298.74921  46.973351      14.478  
6987     0.903  297.18875  47.093819      14.082  
6988     1.031  286.50937  47.163219      14.757  
6989     1.041  294.16489  47.176281      15.385  
6990     1.193  297.00977  47.121021      14.826  

[6991 rows x 21 columns]

In [7]:

data = df[['koi_disposition']]
data_binary_encoded = pd.get_dummies(data, columns=["koi_disposition"])
data_binary_encoded.columns = [["candidate","confirmed","false_positive"]]
y = data_binary_encoded

In [8]:
# Using RandomForestClassifier to find features importance
rf = RandomForestClassifier(n_estimators=200)
rf = rf.fit(Xtemp, y)

In [9]:
# Random Forests in sklearn will automatically calculate feature importance
importances = rf.feature_importances_
importances

array([0.12465095, 0.09081754, 0.12694696, 0.04761473, 0.03965355,
       0.02797612, 0.01679996, 0.0169653 , 0.03710393, 0.03082125,
       0.05151242, 0.08710459, 0.0297809 , 0.03282749, 0.12727127,
       0.02045551, 0.01602031, 0.01701932, 0.02040108, 0.0189583 ,
       0.0192985 ])

In [10]:
# We can sort the features by their importance
sorted(zip(rf.feature_importances_, Xtemp), reverse=True)

[(0.1272712656126526, 'koi_model_snr'),
 (0.12694696456350543, 'koi_fpflag_co'),
 (0.12465095344430142, 'koi_fpflag_nt'),
 (0.09081754400562635, 'koi_fpflag_ss'),
 (0.08710459244243379, 'koi_prad'),
 (0.051512423950009496, 'koi_depth'),
 (0.047614733161617594, 'koi_fpflag_ec'),
 (0.0396535532889173, 'koi_period'),
 (0.037103934459010404, 'koi_impact'),
 (0.03282748654040138, 'koi_insol'),
 (0.030821252897329485, 'koi_duration'),
 (0.029780899802220703, 'koi_teq'),
 (0.027976115842118173, 'koi_time0bk'),
 (0.020455506498321643, 'koi_steff'),
 (0.02040108481869926, 'ra'),
 (0.01929850104771057, 'koi_kepmag'),
 (0.01895830225064709, 'dec'),
 (0.01701932184884426, 'koi_srad'),
 (0.016965295049971008, 'koi_srad'),
 (0.016799958686035493, 'koi_slogg'),
 (0.01602030978962659, 'koi_slogg')]

In [11]:
# Removing features less than 0.26
X = Xtemp.drop(columns=['ra','dec','koi_kepmag','koi_srad','koi_slogg'])
X

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
0                 0              0              0              0   54.418383   
1                 0              1              0              0   19.899140   
2                 0              1              0              0    1.736952   
3                 0              0              0              0    2.525592   
4                 0              0              0              0    4.134435   
...             ...            ...            ...            ...         ...   
6986              0              0              0              1    8.589871   
6987              0              1              1              0    0.527699   
6988              0              0              0              0    1.739849   
6989              0              0              1              0    0.681402   
6990              0              0              1              1    4.856035   

      koi_time0bk  koi_impact  koi_duration  koi_depth  koi_prad  koi_teq  \
0      162.513840       0.586       4.50700      874.8      2.83      443   
1      175.850252       0.969       1.78220    10829.0     14.60      638   
2      170.307565       1.276       2.40641     8079.2     33.46     1395   
3      171.595550       0.701       1.65450      603.3      2.75     1406   
4      172.979370       0.762       3.14020      686.0      2.77     1160   
...           ...         ...           ...        ...       ...      ...   
6986   132.016100       0.765       4.80600       87.7      1.11      929   
6987   131.705093       1.252       3.22210     1579.2     29.35     2088   
6988   133.001270       0.043       3.11400       48.5      0.72     1608   
6989   132.181750       0.147       0.86500      103.6      1.07     2218   
6990   135.993300       0.134       3.07800       76.7      1.05     1266   

      koi_insol  koi_model_snr  koi_steff  
0          9.11           25.8       5455  
1         39.30           76.3       5853  
2        891.96          505.6       5805  
3        926.16           40.9       6031  
4        427.65           40.2       6046  
...         ...            ...        ...  
6986     176.40            8.4       5638  
6987    4500.53          453.3       5638  
6988    1585.81           10.6       6119  
6989    5713.41           12.3       6173  
6990     607.42            8.2       6469  

[6991 rows x 14 columns]

In [27]:
# Set features. This will also be used as your x values.
selected_features = X.columns
selected_features

Index(['koi_fpflag_nt', 'koi_fpflag_ss', 'koi_fpflag_co', 'koi_fpflag_ec',
       'koi_period', 'koi_time0bk', 'koi_impact', 'koi_duration', 'koi_depth',
       'koi_prad', 'koi_teq', 'koi_insol', 'koi_model_snr', 'koi_steff'],
      dtype='object')

In [28]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [29]:
X_train.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
6122              0              0              0              0    6.768901   
6370              0              1              0              1    0.733726   
2879              1              0              0              0    7.652707   
107               0              0              0              0    7.953547   
29                0              0              0              0    4.959319   

      koi_time0bk  koi_impact  koi_duration  koi_depth  koi_prad  koi_teq  \
6122   133.077240       0.150       3.61600      123.1      1.24     1017   
6370   132.020050       0.291       2.30900      114.6      0.86     1867   
2879   134.460380       0.970      79.89690      641.1      3.21      989   
107    174.662240       0.300       2.63120      875.4      2.25      696   
29     172.258529       0.831       2.22739     9802.0     12.21     1103   

      koi_insol  koi_model_snr  koi_steff  
6122     253.30           10.8       5737  
6370    2891.64           13.8       5855  
2879     226.81          254.3       6328  
107       55.37           38.4       4768  
29       349.40          696.5       5712

In [15]:
# Scale your data
from sklearn.preprocessing import StandardScaler
X_scaler = StandardScaler().fit(X_train)
y_scaler = StandardScaler().fit(y_train)

In [16]:
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)
y_train_scaled = y_scaler.transform(y_train)
y_test_scaled = y_scaler.transform(y_test)

In [17]:
model = LinearRegression()
model.fit(X_train_scaled, y_train_scaled)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [18]:
print(f"Training Data Score: {model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {model.score(X_test_scaled, y_test)}")

Training Data Score: -0.7347172338431954
Testing Data Score: -0.7217033280694909


In [19]:
# predictions
predictions = model.predict(X_test_scaled)


In [20]:
MSE = mean_squared_error(y_test_scaled, predictions)
r2 = model.score(X_test_scaled, y_test_scaled)
print(f"MSE: {MSE}, R2: {r2}")

MSE: 0.5608096994436649, R2: 0.4467881931069412


In [21]:
# LASSO model
# Note: Use an alpha of .01 when creating the model for this activity
lasso = Lasso(alpha=.01).fit(X_train_scaled, y_train_scaled)

predictions = lasso.predict(X_test_scaled)

MSE = mean_squared_error(y_test_scaled, predictions)
r2 = lasso.score(X_test_scaled, y_test_scaled)
print(f"MSE: {MSE}, R2: {r2}")

MSE: 0.5605954827898855, R2: 0.4469995075371685


In [22]:
# Ridge model
# Note: Use an alpha of .01 when creating the model for this activity
ridge = Ridge(alpha=.01).fit(X_train_scaled, y_train_scaled)

predictions = ridge.predict(X_test_scaled)

MSE = mean_squared_error(y_test_scaled, predictions)
r2 = ridge.score(X_test_scaled, y_test_scaled)

print(f"MSE: {MSE}, R2: {r2}")

MSE: 0.5608097262746679, R2: 0.4467881666394466


In [23]:
# ElasticNet model
# Note: Use an alpha of .01 when creating the model for this activity

elasticnet = ElasticNet(alpha=.01).fit(X_train_scaled, y_train_scaled)
predictions = elasticnet.predict(X_test_scaled)

MSE = mean_squared_error(y_test_scaled, predictions)
r2 = elasticnet.score(X_test_scaled, y_test_scaled)

print(f"MSE: {MSE}, R2: {r2}")

MSE: 0.5605503626540483, R2: 0.4470440163819665


In [24]:
# Create and score a decision tree classifier
from sklearn import tree
clf = tree.DecisionTreeClassifier()
clf = clf.fit(X_train, y_train)
clf.score(X_test, y_test)

0.868421052631579

In [25]:
# Create and score a Random Forest classifier
rf = RandomForestClassifier(n_estimators=200)
rf = rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.8918764302059496

In [26]:
f = RandomForestClassifier(n_estimators=500)
rf = rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.8947368421052632

Decision Tree Classifier (86.84%) and Random Forest Classifier (89.18%) scores are better than other models.